### PR TITLE
mgr/dashboard: carbon tree component

### PR DESCRIPTION
**Pull request by ivoalmeida**
Replaces the deprecated npm package '@circlon/angular-tree-component' by Carbon Tree component.

Fixes: https://tracker.ceph.com/issues/68249

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [x] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

image
image
image

### CODE REVIEW

File: src/pybind/mgr/dashboard/frontend/package.json [hunk 1]
```diff
@@ -64,7 +63,7 @@
     "@types/file-saver": "2.0.1",
     "async-mutex": "0.2.4",
     "bootstrap": "5.2.3",
-    "carbon-components-angular": "5.25.1",
+    "carbon-components-angular": "5.45.1",
```
[the remark this answers is not in the document]
Comment (ivoalmeida) [reply src/pybind/mgr/dashboard/frontend/package.json]: PTAL https://github.com/ceph/ceph/pull/60626

**Comment by nizamial09** [comment]
Looks like the unit tests are failing: https://jenkins.ceph.com/job/ceph-pull-requests/146280/

**Comment by ivoalmeida** [comment]
> Looks like the unit tests are failing: https://jenkins.ceph.com/job/ceph-pull-requests/146280/

Hi @nizamial09 

Tests are passing now.

**Comment by nizamial09** [comment]
Thank you @ivoalmeida. Looks great and sorry for the delay in testing this. Pointing out some things i saw while testing.

1. CephFS directories

If I select a directory and create a snapshot on it, I am not seeing the value there in the table immediately. I have to go to another tree entry and come back to see the table gets updated. It happens with the quota as well. It could be an unrelated issue but I remember it used to work when I was converting the forms to carbon.

https://github.com/user-attachments/assets/938b4042-a74b-4b1c-a241-863653d6b2bf

2. Multi-site topology viewer
I can't seem to do the actions there. Edit opens up the modal but its broken. I can click the Delete action but it doesn't do anything. I saw the below in console log

```
Uncaught TypeError: Cannot read properties of undefined (reading 'data')
    at RgwMultisiteZoneFormComponent.ngOnInit (rgw-multisite-zone-form.component.ts:184:36)
    at callHook (core.mjs:2434:22)
    at callHooks (core.mjs:2403:17)
    at executeInitAndCheckHooks (core.mjs:2354:9)
    at refreshView (core.mjs:10341:21)
```


https://github.com/user-attachments/assets/29291a67-5bf7-47b5-90d6-1c6d03bf8b69


Incase you want to see more entries in the multisite view, you can [simply do this](https://github.com/rhcs-dashboard/ceph-dev?tab=readme-ov-file#rgw-multi-site). Just setting the `RGW_MULTISITE=1` in .env file and starting the cluster normally `dc up -d ceph` should be fine for the basic testing of the topology viewer. I don't think you'll need to start 2 clusters there.

**Comment by ivoalmeida** [comment]
> Thank you @ivoalmeida. Looks great and sorry for the delay in testing this. Pointing out some things i saw while testing.
> 
> 1. CephFS directories
> 
> If I select a directory and create a snapshot on it, I am not seeing the value there in the table immediately. I have to go to another tree entry and come back to see the table gets updated. It happens with the quota as well. It could be an unrelated issue but I remember it used to work when I was converting the forms to carbon.
> 
>  blob.5.mp4 
> 2. Multi-site topology viewer
>    I can't seem to do the actions there. Edit opens up the modal but its broken. I can click the Delete action but it doesn't do anything. I saw the below in console log
> 
> ```
> Uncaught TypeError: Cannot read properties of undefined (reading 'data')
>     at RgwMultisiteZoneFormComponent.ngOnInit (rgw-multisite-zone-form.component.ts:184:36)
>     at callHook (core.mjs:2434:22)
>     at callHooks (core.mjs:2403:17)
>     at executeInitAndCheckHooks (core.mjs:2354:9)
>     at refreshView (core.mjs:10341:21)
> ```
> 
>  blob.4.mp4 
> Incase you want to see more entries in the multisite view, you can [simply do this](https://github.com/rhcs-dashboard/ceph-dev?tab=readme-ov-file#rgw-multi-site). Just setting the `RGW_MULTISITE=1` in .env file and starting the cluster normally `dc up -d ceph` should be fine for the basic testing of the topology viewer. I don't think you'll need to start 2 clusters there.

Thanks @nizamial09 
I think I have addressed the issues you spotted. PTAL

**Comment by nizamial09** [comment]
tested again, thanks for addressing the other comments.

I found 2 issues again but could be minor ones. Apart from that it looks good to me.

- when we have many items in the tree than it should in the view, shouldn't it have a scroll bar?
image

- I created a zonegroup on a realm, and whenever i refresh the topology viewer page (sometimes i just have to go back to other pages and come back), and it shows this ExpressionChanged error in console

image

**Comment by ivoalmeida** [comment]
> I can see the scroll bar in directories page and also the zonegroup creation issue is gone.
> 
> But I don't see scrollbar for multi-site view and I feel like the multisite topology viewer is stuttery when I scroll the page down and up. Also I seen the page not loading anything when I move away from the page and comeback. I need to refresh the page.
> 
>  blob.7.mp4 
> And I saw that iSCSI topology viewer doesn't show the item name. image

Hi @nizamial09 thanks for spotting these issues. The old tree component did a lot of heavy lifting to optimise performance. I did see the CPU utilisation was really high and I fixed that by changing the change detection strategy to OnPush and detect changes only when new data is updated. Now CPU utilisation is very low which fantastic. the other issue was the data structure of context had to be put inside a data property instead of spreading them all in the root of each node element. This was fixed.

**Comment by nizamial09** [comment]
@ivoalmeida thank you for the explanation and the fix. Its a lot better now :partying_face: Although, scroll bar is still not there in multi-site topology viewer and the name is not there in iSCSI topology viewer either.

**Comment by nizamial09** [comment]
jenkins test make check

**Comment by nizamial09** [comment]
jenkins test make check arm64

**Comment by nizamial09** [comment]
jenkins test api

**Review by nizamial09 — APPROVED** [review]
works perfectly now, thank you :partying_face:

**Comment by ivoalmeida** [comment]
> @ivoalmeida unit tests [are failing btw](https://jenkins.ceph.com/job/ceph-pull-requests/147463/testReport/junit/projectroot.src.test/mgr/mgr_dashboard_frontend_unittests/)

They should be passing now. Thanks

**Comment by ivoalmeida** [comment]
jenkins test api

**Pull request merged**